### DISCUSSION TITLE
Exporting to html-wasm with imported modules

### DISCUSSION THREAD

**Discussion by q-wertz** [marimo-team/marimo#3158] [Q&A] [answered]
Hello,

not sure if blind or misunderstanding something, but I fail when trying to export a marimo notebook that imports a python file with some extra functions to html-wasm.

Project layout:
```
.
└── Project directory/
    ├── helpfunctions/
    │   ├── __init__.py
    │   └── helpfunctions.py
    └── marimo_nb.py
```

In the marimo notebook `marimo_nb.py` I import helpfunctions using `from helpfunctions import helpfunctions`

If I convert now to html-wasm and then try to use the generated stuff 
```shell
$ marimo export html-wasm marimo_nb.py -o build_marimo --mode run
$ cd build_marimo
$ python -m http.server
```

I get an `ModuleNotFoundError` error 
```
ModuleNotFoundError

    This cell raised an exception: ModuleNotFoundError('No module named 'helpfunctions'')
    See the console area for a traceback.
```

Is this an error on my side (I mean I can also add the helpfunctions into the marimo notebook but it clutters things up) or should marimo export those functions (according to the [documentation](https://docs.marimo.io/guides/exporting.html#export-to-wasm-powered-html) "The exported HTML file will run your notebook using WebAssembly, making it completely self-contained and executable in the browser.")

**Comment by mscolnick** [comment, marked as answer]
hey @q-wertz - we should update our documentation, but it only packages your notebook.py and does not traverse for other files in your directory. this is something we could potentially support in future, but for now you can:

1. include all your files in the notebook.py (or write a small script to do that beforehand)
2. publish your helper functions to a PyPi or a private registry and marimo can install them (this is probably too much friction for a few helper functions, but better for larger libraries)

**Reply by q-wertz** [reply]
Thanks for the fast reply 😊
Would be awesome if this would be possible in the future.
For the moment I added the functions to the notebook.

Btw: really nice feature 😊

**Reply by gurvinder-dhillon** [reply]
Has this been implemented or can be tracked somewhere ?

**Comment by asmunder** [comment]
Just in case anyone else is looking for ways to do this, the following bash script may be useful.
It will take your notebook file, look for any lines on the form `#inline foo.py`, and insert files with those names from the directory specified in the `pkgdir` variable.

The script does some basic level of error checking, and asks the user to confirm the list of files to be inserted, but **please read the script it before using it to make sure it fits your use case**. The only argument accepted by the script is `-f` which means to skip the confirmation prompt, useful e.g. if this is used in a github pipeline or something.

```bash
#!/bin/bash

file="notebook.py"
pkgdir="mypackage"

# Check that the specified file and directory exists
if [[ ! -f "$file" ]]; then
  echo "Error: your must specify in the script an exisiting file to process"
  exit 1
fi
if  [[ ! -d $pkgdir ]]; then
  echo "Error: your must specify in the script a directory which has the source files"
  echo '  You can specify "." for the current directory'
  exit 1
fi

# First we read the file and get a list of what to insert where
linenum=0
declare -a linenums
declare -a inserts

while IFS= read -r line; do
  ((linenum ++))
  if [[ "$line" == "#inline "* ]]; then
    linenums+=($linenum)
    linearr=($line)
    inserts+=(${linearr[1]})
  fi
done < "$file"

# If the "-f" flag has not been given, ask the user if they want to proceed
if [[ $1 == "-f" ]]; then
  response="y"
else
  echo "The program will perform the following inlining:"
  echo "Files to insert:" "${inserts[@]}"
  echo "After line numbers:" "${linenums[@]}"

  read -r -p "Do you want to proceed? [y/N] " response
fi

if [[ "$response" =~ ^(yes|y)$ ]]; then
    # Do the actual processing. First make a copy of our file, restore it at the end.
    cp $file $file.bak

    # Loop over filenames to insert
    for ins in "${inserts[@]}"; do
      if [[ ! -f $pkgdir/$ins ]]; then
        echo "Error: could not find $pkgdir/$ins, skipping..."
      else
        # This inserts the file
        sed -i -e "/#inline $ins/r $pkgdir/$ins" $file
        # This changes the comment string, so you could run this multiple times
        sed -i "/#inline $ins/#Have inlined $ins" $file
      fi
    done

    # Move result to output_ file and restore copy of original file
    mv $file output_$file
    mv $file.bak $file

else
    echo "Cancelled the process."
    exit 1
fi
```
